### ISSUE TITLE
Sink should generate log URLs for Pods

### ISSUE THREAD

**Issue by ggallen** [kubearchive/kubearchive#538]
Whenever a `Pod` is archived by KubeArchive, the sink should generate log URLs for the `Pod` if logging is configured.

### Configuration

The sink should check the KubeArchive configuration `ConfigMap` to see if logging is enabled. If it is, it should read the logging `ConfigMap` specified there and create a map for generating log URLs from the contents of that `ConfigMap`.  That map should contain key-value pairs for everything in the `ConfigMap`.  Values starting with the prefix "cel:" should be compiled as CEL expressions, and the value for those keys containing CEL expressions should be the CEL program that resulted from the compilation.  All other values are copied to the map as strings.

The logging `ConfigMap` needs to be watched for changes, and the map regenerated when changes to it are detected.

This map should be generated only once and reused for all log URL generation.

### Log URL Generation

If a `Pod` is archived and log URL generation is enabled as described above, the sink needs to generate the log URLs for that `Pod` and store them in the `log_url` table. These steps should be done:

1. All existing entries in the `log_url` table for the`Pod` should be deleted.
1. Log URLs are generated by calling `log_url.log_url.GenerateLogURLs, passing in the map of key-value pairs to be used for substitution.
1. For each log URL returned by the previous step, add a new record to the `log_url` table for this `Pod`.

### Acceptance Criteria

1. The sink detects and watches for logging configuration and handles it appropriately.
1. Log URLs are generated for `Pods` and stored in the table. 

### References

- [Logging design document](https://github.com/kubearchive/kubearchive/blob/main/docs/modules/ROOT/pages/design/logging-integration.adoc)

**Comment by skoved** [comment]
[Logging design doc web](https://kubearchive.github.io/kubearchive/main/design/logging-integration.html)

**Comment by skoved** [comment]
the sink needs to get the logging configuration from the `kubearchive-config` ConfigMap. This has me thinking about how to we should handle it. We have a few options:

1. `Get` the ConfigMap each time a pod is archived. **Pro** is that we don't need additional watch logic. **Con** is that if the `kubearchive-config` ConfigMap doesn't change often, we will make more http requests than if we `watch`
2. `Watch` the ConfigMap. There are 2 versions of this but both have a **pro** that if the logging configuration doesn't change frequently, less http requests will be made to retrieve the logging configuration. Both have a `con` that they will add more logic to a separate goroutine.
  a. `kubearchive-config` is added to the same `watch` as `sink-filters`. A `watch` can return updates for all or multiple objects of a specific kind. The events that are received contain the actual object that changed, so `event.Object.GetName()` can be used to determine which ConfigMap the event was for. **Pro** is we don't need an additional goroutine. **Con** is we will need additional logic in the `watch` logic to demultiplex the updates.
  b. `kubearchive-config` gets its own `watch` in a separate goroutine. **Pro** is we don't have logic for updating filters and logging configuration tangled up together. **Con** is the sink will have a 2nd goroutine running outside of the main routine.

**Comment by rh-hemartin** [comment]
Doing a Get for every Pod update (option 1) will hammer the Kubernetes API, even if they cache the answer. Having a watch coupled with `sink-filters` would make it a little bit confusing. I think having a watch in its own goroutine (option 2b) is what makes the most sense. Watchers match the use case and goroutines are cheap in go.

**Comment by ggallen** [comment]
Option a is not viable.

Option 2b makes the most sense. And it's the one I was thinking of when I created this issue.

**Comment by skoved** [comment]
After discussion with Greg earlier today, we're going to attempt a v1 of this with a third option that was not listed previously. When the sink starts it will reach out and look for the logging configuration configmap, do a get on that configmap and load the configuration. if the logging configuration changes, the sink will need to be restarted. Once this is working, we can look into potential performance impacts on doing another watch as well has creating a sig handler to reload the logging configuration (if there is an easy way to send a specified signal from Kubernetes). I've seen linux daemons use `SIGHUP` as way to make the daemon reload its config file.

**Comment by skoved** [comment]
> . When the sink starts it will reach out and look for the logging configuration configmap, do a get on that configmap and load the configuration.

being more specific about what the sink will do at start up with respect to the logging config

1. The sink checks for and reads the kubearchive-config ConfigMap.
2. If a logging ConfigMap is specified in the kubearchive-config ConfigMap, it is read.
3. If the logging ConfigMap is read, it is used to create a map to be passed to GenerateLogURLs, where the keys are the keys in the ConfigMap and the values are either the CEL program (if the value represents a CEL expression, i.e. starts with "cel:") or the value as-is as a string.

**Comment by skoved** [comment]
I noticed a few of the things about the `public.log_url` table

1. There is no primary key
2. uuid is not marked a foreign key even though it should be (`public.resources` uses `uuid` as its primary key)
3. if user provides an `archiveWhen` CEL expresion in their `KubeArchiveConfig`, the sink may archive the pod more than once (depends on the CEL expression provided). In this senario, `public.log_url` will end up with multiple rows with the same values

I propose the following

1. make `url` the primary key
2. mark uuid as a foreign key
3. after making the change for [1], the sql query that the sink uses to write log urls can use `ON CONFLICT` to prevent writes when a `url` has already been written to the table

If there is agreement I'll create a separate issue to address these problems

**Comment by ggallen** [comment]
> There is no primary key

Yes, there isn't. And I don't think there's a need for one.  I don't think `url` should be  a primary key. In general, long string values aren't recommended to be primary key fields.

> uuid is not marked a foreign key even though it should be (public.resources uses uuid as its primary key)

But it is. From the schema:
```
ALTER TABLE ONLY public.log_url
    ADD CONSTRAINT fk_uuid_resource FOREIGN KEY (uuid) REFERENCES public.resource(uuid);
```
> if user provides an archiveWhen CEL expresion in their KubeArchiveConfig, the sink may archive the pod more than once (depends on the CEL expression provided). In this senario, public.log_url will end up with multiple rows with the same values

This is why I suggested that the sink should *delete* existing logs for the `uuid` before adding new ones.

**Comment by skoved** [comment]
I missed the bit with the foreign key. I'll give it a shot with deleting all the URLs with the `uuid` before inserting. I'll see if I can benchmark the delete operation when there are lots of logs in the database.

**Comment by ggallen** [comment]
>  I'll see if I can benchmark the delete operation when there are lots of logs in the database.

Please do this as a separate issue after you have implemented the changes. If we have to we can change it later, but we need to get logging implemented.

I don't expect the deletes to be a performance issue.  The delete should be done by `uuid`, and there is an index on that field.

**Comment by skoved** [comment]
> I don't expect the deletes to be a performance issue. The delete should be done by uuid, and there is an index on that field.

I think I missed the index for `uuid` in the sql file for creating the database.

**Comment by skoved** [comment]
@ggallen we had a discussion a couple weeks ago about my concern with create watchers in the sink for all of the different `ConfigMap`s. Particularly that logging would require creating 2 watchers. One to watch `kubearchive-config` and an other that watches the logging `ConfigMap` referenced in `kubearchive-config`. We could reduce the number of `ConfigMaps` the sink needs to watch by using a `ConfigMap`. Here's my idea:

1. User creates some number of logging `ConfigMap`s
2. User creates the `kubearchive-config` `ConfigMap`
3. The operator (or the user if you don't want the operator to do this) adds a label (`kubearchive-logging-config: active` for example) to the `ConfigMap` listed in `logging-configmap-name`
4. If that field changes the operator (or the user) removes `kubearchive-logging-config: active` from the previous logging `ConfigMap` and applies that label to the new logging `ConfigMap`

This would let the sink listen on for ConfigMaps wh

**Comment by ggallen** [comment]
@skoved, it doesn't look like you finished your last comment.

**Comment by skoved** [comment]
sorry, updated my last comment

**Comment by ggallen** [comment]
There are too many manual steps there all around.  The user should just be able to change the kubearchive-config `ConfigMap` to use a new name and it should all just happen.

How about using something like this or something similar?

https://github.com/stakater/Reloader

**Comment by ggallen** [comment]
> This would let the sink create a watcher and give it the following label selector kubearchive-logging-config = active.

This still doesn't help with the "kubearchive-config" `ConfigMap`. I can imagine we'll have more setting in this `ConfigMap` eventually, and if it changes we'll need to restart things.

**Comment by rh-hemartin** [comment]
Having a watcher on `kubearchive-config` for each KubeArchive that stops the app sounds like a nice solution. Kubernetes will spin up a new pod and that will get the new configuration at startup. We also avoid adding a new dependency.

**Comment by ggallen** [comment]
> Having a watcher on `kubearchive-config` for each KubeArchive that stops the app sounds like a nice solution. Kubernetes will spin up a new pod and that will get the new configuration at startup. We also avoid adding a new dependency.

But that's exactly what @skoved didn't want to do, if I understand correctly.

**Comment by skoved** [comment]
> > Having a watcher on `kubearchive-config` for each KubeArchive that stops the app sounds like a nice solution. Kubernetes will spin up a new pod and that will get the new configuration at startup. We also avoid adding a new dependency.
> 
> But that's exactly what @skoved didn't want to do, if I understand correctly.

What I'm trying to avoid is the recursive watch that is required for logging. As it stands now, the sink would need to do the following

1. watch the`kubearchive-config` ConfigMap
2. If the key `logging-configmap-name` is added or changed, a second watch needs to be started for the logging ConfigMap. If the key `logging-configmap-name` is removed, the watch on the logging ConfigMap must be stopped. If the `kubearchive-config` ConfigMap is deleted, both watches must be stopped
3. Events from the watch on the logging ConfigMap are used to update the logging data in the sink

I'm not against watching `kubearchive-config` or the logging ConfigMap on their own, but I'm worried about the complexity introduced by the watch for the logging ConfigMap being dependent and being managed by the watch on `kubearchive-config`. My personal feeling is that if I implement what I described above, I will be the only person that understands how the code works and it will be complicated enough that nobody else will want to touch it

**Comment by maruiz93** [comment]
It's not the first time I looked for the feature "automatic rollout deployment on configmap changes" so I would like to paste here some links just for reference:

* There is an [issue opened](https://github.com/kubernetes/kubernetes/issues/22368) asking for this of more than 8 years, there is a lot of interest but I guess it's not that simple. 
* There is an already pointed controller [Reloader](https://github.com/stakater/Reloader) that implies having one extra dependency
* Just for reference, if Openshift is used, it provides a `DeploymentConfig` with a `ConfigChange` [trigger](https://docs.openshift.com/container-platform/3.11/dev_guide/deployments/basic_deployment_operations.html#config-change-trigger) that allows automatic rollouts when config changes.
* Update the livenessProbe to include a check of the hash of the mounted configmap and check if it has changed. This can be done in a script before calling the /healz endpoint to do both checks

Even if we choose to postpone the automation of this process, I believe it's worth considering the livenessProbe approach at some point.
